libnetdata/json/json.c

```diff
@@ -0,0 +1,294 @@
+	ne.original_string = &js[t[start].start];
+
+	memcpy(&ne, e, sizeof(JSON_ENTRY));
+	ne.type = JSON_ARRAY;
+	ne.data.items = t[start].size;
+	ne.callback_function = NULL;
+	strcpy(ne.name, "");
+	strcpy(ne.fullname, ne.name);
+	if(e->callback_function) e->callback_function(&ne);
+	js[t[start].end] = old;
+
+	size_t i, init = start, size = t[start].size;
+
+	start++;
+	for(i = 0; i < size ; i++) {
+		ne.pos = i;
+		if (!e->name || !e->fullname || strlen(e->name) > JSON_NAME_LEN  - 24 || strlen(e->fullname) > JSON_FULLNAME_LEN -24) {
+		    info("JSON: JSON walk_array ignoring element with name:%s fullname:%s",e->name, e->fullname);
+		    continue;
+		}
+		sprintf(ne.name, "%s[%lu]", e->name, i);
+		sprintf(ne.fullname, "%s[%lu]", e->fullname, i);
+
+		switch(t[start].type) {
+			case JSMN_PRIMITIVE:
+				start += json_walk_primitive(js, t, start, &ne);
+				break;
+
+			case JSMN_OBJECT:
+				start += json_walk_object(js, t, nest + 1, start, &ne);
+				break;
+
+			case JSMN_ARRAY:
+				start += json_walk_array(js, t, nest + 1, start, &ne);
+				break;
+
+			case JSMN_STRING:
+				start += json_walk_string(js, t, start, &ne);
+				break;
+		}
+	}
+	return start - init;
+}
+
+size_t json_walk_object(char *js, jsmntok_t *t, size_t nest, size_t start, JSON_ENTRY *e)
+{
+	JSON_ENTRY ne;
+
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	ne.original_string = &js[t[start].start];
+	strcpy(ne.name, "");
+	strcpy(ne.fullname, ne.name);
+	memcpy(&ne, e, sizeof(JSON_ENTRY));
+	ne.type = JSON_OBJECT;
+	ne.callback_function = NULL;
+	if(e->callback_function) e->callback_function(&ne);
+	js[t[start].end] = old;
+
+	int key = 1;
+	size_t i, init = start, size = t[start].size;
+
+	start++;
+	for(i = 0; i < size ; i++) {
+		switch(t[start].type) {
+			case JSMN_PRIMITIVE:
+				start += json_walk_primitive(js, t, start, &ne);
+				key = 1;
+				break;
+
+			case JSMN_OBJECT:
+				start += json_walk_object(js, t, nest + 1, start, &ne);
+				key = 1;
+				break;
+
+			case JSMN_ARRAY:
+				start += json_walk_array(js, t, nest + 1, start, &ne);
+				key = 1;
+				break;
+
+			case JSMN_STRING:
+			default:
+				if(key) {
+					int len = t[start].end - t[start].start;
+					if (unlikely(len>JSON_NAME_LEN)) len=JSON_NAME_LEN;
+					strncpy(ne.name, &js[t[start].start], len);
+					ne.name[len] = '\0';
+					len=strlen(e->fullname) + strlen(e->fullname[0]?".":"") + strlen(ne.name);
+					char *c = mallocz((len+1)*sizeof(char));
+					sprintf(c,"%s%s%s", e->fullname, e->fullname[0]?".":"", ne.name);
+					if (unlikely(len>JSON_FULLNAME_LEN)) len=JSON_FULLNAME_LEN;
+					strncpy(ne.fullname, c, len);
+					freez(c);
+					start++;
+					key = 0;
+				}
+				else {
+					start += json_walk_string(js, t, start, &ne);
+					key = 1;
+				}
+				break;
+		}
+	}
+	return start - init;
+}
+
+size_t json_walk_tree(char *js, jsmntok_t *t, void *callback_data, int (*callback_function)(struct json_entry *))
+{
+	JSON_ENTRY e;
+	strcpy(e.name, "");
```

Comment:
`ne.name[0] = '\0'` has the same effect and it is way faster.

Reply:
Done